File: calc.pas
unit IntegerTerm;

interface

uses SysUtils;

type
  VarRec = record
    Name: string;
    Wert: variant;
  end;
  VarArray = array of VarRec;
  PVarArray = ^VarArray;

  TIntTerm = class
    constructor Create();
  private
    { Private-Deklarationen }
    pVari: PVarArray;
    FehlerPos: integer;
    FehlerLaenge: integer;
    FehlerText: string;
    procedure FindeOperation(Term, Op: string; pPos: pinteger; p: integer);
    function  TermZerlegung(Term: string; p: integer): integer;
    procedure SetFehlerMeldung(s: string; p: integer; l: integer);
  public
    { Public-Deklarationen }
    function  TermLoesen(Term: string; pVariablen: PVarArray): integer;
    function GetFehler(FPos, FLaenge: pinteger): integer;
    function GetFehlerText: string;
  end;

const
  // Zeichen bzw. Symbole für Rechenoperation
  //    << Bits links schieben, >> Bits rechts schieben
  //    & AND, | OR, ~ XOR
  OpStr: array[0..8] of string = ('+','-','*','/','>>','<<','&','|','~');

implementation

constructor TIntTerm.Create;
begin
  pVari := nil;
  FehlerPos := 0;
  FehlerLaenge := 0;
  FehlerText := '';
end;

function TIntTerm.TermLoesen(Term: string; pVariablen: PVarArray): integer;
begin
  FehlerPos := 0;
  FehlerText := '';
  pVari := pVariablen;
  result := TermZerlegung(Term, 1);
end;

// -----------------------------------------------------------------------------
//  FindeOperation
//
//  Sucht die Rechenopertion, und gibt die Position zurück. In Klammern wird
//  nicht gesucht.
//
//  Term : Term, der nach einer Operation/Verknüpfung durchsucht werden soll
//  Op : Operation/Verknüpfung nach der in Term gesucht werden soll
//  pPos : Rückgabe der Position an der die Operation gefunden wurde
//  p : Zeichenposition im gesamt Term. Dient der Fehlerausgabe
// -----------------------------------------------------------------------------
procedure TIntTerm.FindeOperation(Term: string; Op: string; pPos: pinteger; p: integer);
var
  n: integer;     // Zeichenzähler
  k: integer;     // Klammerzähler, '(' plus 1 ; ')' minus 1
  kPos: integer;  // Merker für Position der ersten Klammer (Fehlerausgabe)
begin
  kPos := 0;
  k := 0;
  for n := 1 to length(Term) do
  begin
    if term[n] = '(' then
    begin
      if k = 0 then
        kPos := n;
      inc(k);
    end;
    if term[n] = ')' then
    begin
      dec(k);
      if k < 0 then
      begin
        SetFehlerMeldung('Klammer nicht geöffnet.', p+n-1, 1);
        exit;
      end;
    end;
    // if (k = 0) and (term[n] = Op) then
    if (k = 0) and (copy(term, n, length(Op)) = Op) then
    begin
      pPos^ := n;
      exit;
    end;
  end;
  pPos^ := 0;
  if k > 0 then
    SetFehlerMeldung('Öffnende Klammer gefunden, die nicht geschlossen wird.', p+kPos-1, 1);
end;
// ^^ FindeOperation ^^^^^^^^^^^^^^^^^^^^^^^^^^

// -----------------------------------------------------------------------------
// TermZerlegung
//
// Teilt den Term in seine Bestandteile auf
//
// Term : String des Term der entschlüsselt werden soll
// p : Zeichenposition im gesamt Term. Dient der Fehlerausgabe
// -----------------------------------------------------------------------------
function TIntTerm.TermZerlegung(Term: string; p: integer): integer;
var
  i: integer;
  n: integer;
  o1: integer;
  o2: integer;
  TempStr: string;
begin
  result := 0;

  if FehlerPos > 0 then
  begin
    result := 1;
    exit;
  end;

  p := p + length(Term) - length(TrimLeft(Term));
  Term := trim(Term);

  i := 0;
  while (i <= high(OpStr)) do
  begin
    FindeOperation(Term, OpStr[i], @n, p);
    if n > 0 then
    begin
      o1 := TermZerlegung(copy(Term, 1, n-1), p);
      o2 := TermZerlegung(copy(Term, n+length(OpStr[i]), length(Term)-n), p+n+length(OpStr[i])-1);
      case i of
        0: result := o1 + o2;
        1: result := o1 - o2;
        2: result := o1 * o2;
        3: result := o1 div o2;
        4: result := o1 shr o2;
        5: result := o1 shl o2;
        6: result := o1 and o2;
        7: result := o1 or o2;
        8: result := o1 xor o2;
      end;
      exit;
    end;
    inc(i);
  end;

  // Ist Term in Klammern
  if Term[1] = '(' then
  begin
    if Term[length(Term)] = ')' then
    begin
      result := TermZerlegung(trim(copy(Term, 2, length(Term)-2)), p+1);
      exit;
    end;
  end;

  // Auf Variable oder Funktion prüfen
  if Term[1] in ['a'..'z','A'..'Z','_'] then
  begin
    i := 2;
    while ((i <= length(Term)) and (Term[1] in ['a'..'z','A'..'Z','_','0'..'9'])) do
    begin
      inc(i);
    end;
    TempStr := copy(Term, 1, i-1);

    // Auf Variablen prüfen
    if assigned(pVari) then
    begin
      for n := low(pVari^) to high(pVari^) do
      begin
        if lowercase(pVari^[n].Name) = lowercase(TempStr) then
        begin
          result := pVari^[n].Wert;
          exit;
        end;
      end;
    end;
    SetFehlerMeldung('Syntax Fehler, Variable "'+TempStr+'" unbekannt.', p, length(TempStr));

    // Auf Funktionen prüfen (z.B. Not, sin, cos)
  end;

  // Wenn Term eine Zahl ist
  if Term[1] in ['0'..'9'] then
  begin
    try
      result := StrToInt(Term);
    except
      SetFehlerMeldung('Syntax Fehler, "'+Term+'" ist keine gültige Zahl.', p, length(Term));
    end;
    exit;
  end;
end;
// ^^ TermZerlegung ^^^^^^^^^^^^^^^^^^^^^^^^^^^

procedure TIntTerm.SetFehlerMeldung(s: string; p: integer; l: integer);
begin
  FehlerPos := p;
  FehlerLaenge := l;
  FehlerText := s;
  raise Exception.Create(s);
end;

function TIntTerm.GetFehler(FPos: pinteger; FLaenge: pinteger): integer;
begin
  result := 0;
  FPos^ := FehlerPos;
  FLaenge^ := FehlerLaenge;
  if FehlerPos > 0 then
    result := -1;
end;

function TIntTerm.GetFehlerText(): string;
begin
  result := FehlerText;
end;


{
procedure TForm1.Button1Click(Sender: TObject);
var
  test: VarArray;
begin
  SetLength(test, 4);
  test[0].Name := 'ertz';
  test[0].Wert := 25;
  test[1].Name := 's34';
  test[1].Wert := 3;
  test[2].Name := 'w3tz';
  test[2].Wert := 18;
  test[3].Name := 'w';
  test[3].Wert := 9;
  Memo1.Clear;
  Memo1.Lines.Add(format('Ergebnis = %d',[TermLoesen(Edit1.Text, @test)]));
end;
}

end.

File: calc.c
// Funktionen zum berechnen des Ergebnis

#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <ctype.h>
#include <string.h>
#include "calc.h"

const char *termstart;
int64_t  result;
//uint16_t FehlerPos;     //: integer;
uint16_t FehlerNum = OK;
char FehlerText[MAX_ERROR_LEN] = {0};     //: string;
//void *pVari = 0;     //: PVarArray;

// Zeichen bzw. Symbole für Rechenoperation
//    << Bits links schieben, >> Bits rechts schieben
//    & AND, | OR, ~ XOR
const char *OpStr[OP_NUMBER] = {"+","-","*","/",">>","<<","&","|","~"};      //: array[0..8] of string = ('+','-','*','/','>>','<<','&','|','~');

bool find_operation(const char *pStart, const char *pEnd, const char *Opp, const char **pOpp);
uint16_t term_disassemble(const char *term, const char *term_end, int64_t *subresult);

/** ***************************************************************************
 * Name
 *
 * Discription
 *
 * @param term  Term-String, der berechnet werden soll
 * @return Fehlernummer
 *************************************************************************** */
uint16_t calc_term(const char *term)
{
    result = 0;
    FehlerNum = OK;
    if (!term) {
      FehlerNum = ERR_TERM_EMPTY;
      return FehlerNum;
    }

    if (!strlen(term)) {
      FehlerNum = ERR_TERM_EMPTY;
      return FehlerNum;
    }

    termstart = term;

    printf("Start Term [%s] %p %p \n", term, term, term + strlen(term));

    return term_disassemble(term, term + strlen(term) - 1, &result);
}

/** ***************************************************************************
 * term_disassemble : 
 *
 * Teilt den Term in seine Bestandteile auf
 *
 * @param term      String des Term der entschlüsselt werden soll
 * @param term_end  Zeige auf letztes Zeichen des Terms
 * @return Fehlernummer
 *************************************************************************** */
uint16_t term_disassemble(const char *term, const char *term_end, int64_t *subresult)
{
    const char *pPos = term;         // Zeiger auf Position die verarbeitet wird
    const char *pEnd = term_end;   // Zeiger auf Ende des Terms
    const char *pOpp = NULL;
    int32_t i = 0;      //: integer;
    int64_t left_opp = 0;        //left operant : integer;
    int64_t right_opp = 0;        //right operant : integer;
    int64_t tempresult = 0;

    printf("Sub Term 1 [%ld] [%ld] \n", pPos-termstart, pEnd-termstart);

    if (FehlerNum != OK) {
        return FehlerNum;
    }

    // Trim left (find begin)
    while (isspace(*pPos)) {
        pPos++;
    }

    printf("Sub Term 2 [%ld] [%ld] \n", pPos-termstart, pEnd-termstart);

    // Trim right (find end)
    while(pEnd > pPos) {
        if (!isspace(*pEnd)) {
            break;
        }
        pEnd--;
    }

    printf("Sub Term 3 [%ld] [%ld] \n", pPos-termstart, pEnd-termstart);

    printf("Subterm start[%d] end[%d] schar[%c] echar[%c]\n"
        , (uint16_t)(pPos-termstart), (uint16_t)(pEnd-termstart), (char)*pPos, (char)*pEnd);

    i = 0;
    while (i < OP_NUMBER) {
      if (find_operation(pPos, pEnd, OpStr[i], &pOpp)) {
        printf("Operation erkannt [%s] Pos[%d]\n", OpStr[i], (uint16_t)(pOpp-termstart));
        if (term_disassemble(pPos, pOpp-1, &left_opp)) {
          return FehlerNum;
        }
        if (term_disassemble(pOpp+1, pEnd, &right_opp)) {
          return FehlerNum;
        }

        printf("Rechen Operanten %ld %s %ld\n", left_opp, OpStr[i], right_opp);

        switch (i) {
          case 0: tempresult = left_opp + right_opp; break;
          case 1: tempresult = left_opp - right_opp; break;
          case 2: tempresult = left_opp * right_opp; break;
          case 3: tempresult = left_opp / right_opp; break;
          case 4: tempresult = left_opp >> right_opp; break;
          case 5: tempresult = left_opp << right_opp; break;
          case 6: tempresult = left_opp & right_opp; break;
          case 7: tempresult = left_opp | right_opp; break;
          case 8: tempresult = left_opp ^ right_opp; break;
          default:
              ;
        }
        printf("Zwischenergebnis %ld\n", tempresult);
        *subresult = tempresult;
        return OK;
      }
      i++;
    }

    // Ist Term in Klammern
    if (*pPos == '(') {
        if (*pEnd == ')') {
          printf("Klammerauflösung\n");
          if (term_disassemble(pPos+1, pEnd-1, &tempresult)) {
            return FehlerNum;
          }
          *subresult = tempresult;
          return OK;
        } else {
          FehlerNum = ERR_BRACKET_CLOSE;
          return FehlerNum;
        }
    }

    // Auf Variable oder Funktion prüfen
    if (isalpha(*pPos) || *pPos == '_') {
      printf("Variablenauflösung\n");
      uint16_t i = 0;
      char *pVarEnd;
      pVarEnd = (char*)pPos;
#warning Fehler zu langer Variablenname abfangen              
      while ((pVarEnd <= pEnd) && (isalnum(*pVarEnd) || *pVarEnd == '_')) {
        pVarEnd++;
        i++;
      }
      char varName[64] = {0};
      strncpy(varName, pPos, i-1);
      *pVarEnd = '\0';
      printf("Variable %s erkannt\n", varName);

      snprintf(FehlerText, MAX_ERROR_LEN, "Syntax Fehler, Variable unbekannt.");
      FehlerNum = ERR_UNKNOWN_VARIABLE;
      return FehlerNum;
    }

    // Wenn mit Komma beginnt = 0,xx
    if (*pPos == ',') {

    }

    // Wenn Term eine Zahl ist
    if (isdigit(*pPos)) {
      while(isdigit(*pPos) && pPos <= pEnd) {
        tempresult *= 10;
        tempresult += *pPos & 0x0f;
        pPos++;
      }
      printf("Zahl erkannt %ld\n", tempresult);
      *subresult = tempresult;
    } else {
      FehlerNum = ERR_UNKNOWN;
      return FehlerNum;
    }

    return OK;
}
// ^^ TermZerlegung ^^^^^^^^^^^^^^^^^^^^^^^^^^^

/** ***************************************************************************
 * find_operation
 *
 * Sucht die Rechenopertion, und gibt die Position zurück. In Klammern wird
 * nicht gesucht.
 *
 * @param Term  Term, der nach einer Operation/Verknüpfung durchsucht werden soll
 * @param Opp   Operation/Verknüpfung nach der in Term gesucht werden soll
 * @param pPos  Rückgabe der Position an der die Operation gefunden wurde
 * @return true wenn Rechenoperation gefunden
 *************************************************************************** */
//bool find_operation(const char *subterm, const char *Opp, const char **pPos)
bool find_operation(const char *pPos, const char *pEnd, const char *Opp, const char **pOpp)
{
  uint16_t i = 0;   // Zeichenzähler
  uint16_t k = 0;   // Klammerzähler, '(' plus 1 ; ')' minus 1

  while (pPos <= pEnd) {
    if (*pPos == '(') {
      k++;
    }

    if (*pPos == ')') {
      if (k == 0) {
        FehlerNum = ERR_BRACKET_OPEN;
        snprintf(FehlerText, MAX_ERROR_LEN, "Klammer nicht geöffnet.");
        return false;
      }
      k--;
    }

    if ((k == 0) && !strncmp(pPos, Opp, strlen(Opp))) {
      printf("Find Opp start[%d] end[%d] Opp[%s]\n", (uint16_t)(pPos-termstart), (uint16_t)(pEnd-termstart), Opp);
      *pOpp = pPos;
      return true;
    }
    i++;
    pPos++;
  }

  if (k > 0) {
    FehlerNum = ERR_BRACKET_CLOSE;
    snprintf(FehlerText, MAX_ERROR_LEN, "Geöffnete Klammer gefunden, die nicht geschlossen wird.");
    return false;
  }

  return false;
}
// ^^ find_operation ^^^^^^^^^^^^^^^^^^^^^^^^^^

/** ***************************************************************************
 * Name
 *
 * Discription
 *
 * @param a  Discription
 * @return no
 *************************************************************************** */
int64_t get_calc_result() {
  return result;
}

/** ***************************************************************************
 * Name
 *
 * Discription
 *
 * @param a  Discription
 * @return no
 *************************************************************************** */
int16_t GetFehler() {
  return FehlerNum;
}

/** ***************************************************************************
 * Name
 *
 * Discription
 *
 * @param a  Discription
 * @return no
 *************************************************************************** */
const char *GetFehlerText() {
  return NULL;
}
